lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+        %{
+          state
+          | caller: def_kind in [:defmacro, :defmacrop, :defguard, :defguardp]
+        }
+        |> new_func_vars_scope()
+      else
+        # make module variables accessible if there are unquote fragments in def body
+        %{
+          state
+          | caller: def_kind in [:defmacro, :defmacrop, :defguard, :defguardp]
+        }
+        |> new_vars_scope()
+      end
+
+    # no need to reset versioned_vars - we never update it
+    env_for_expand = %{env | function: {name, arity}}
+
+    # expand defaults and pass args without defaults to expand_args
+    {args_no_defaults, args, state} =
+      expand_defaults(args, state, %{env_for_expand | context: nil}, [], [])
+
+    # based on :elixir_clauses.def
+    {e_args_no_defaults, state, env_for_expand} =
+      expand_args(args_no_defaults, %{state | prematch: {%{}, 0, :none}}, %{
+        env_for_expand
+        | context: :match
+      })
+
+    args =
+      Enum.zip(args, e_args_no_defaults)
+      |> Enum.map(fn
+        {{:"\\\\", meta, [_, expanded_default]}, expanded_arg} ->
+          {:"\\\\", meta, [expanded_arg, expanded_default]}
+
+        {_, expanded_arg} ->
+          expanded_arg
+      end)
+
+    {e_guard, state, env_for_expand} =
+      __MODULE__.Clauses.guard(
+        guards,
+        %{state | prematch: :raise},
+        %{env_for_expand | context: :guard}
+      )
+
+    type_info = Guard.type_information_from_guards(e_guard)
+
+    state = merge_inferred_types(state, type_info)
+
+    env_for_expand = %{env_for_expand | context: nil}
+
+    state =
+      state
+      |> add_current_env_to_line(meta, env_for_expand)
+      |> add_func_to_index(
+        env,
+        name,
+        args,
+        extract_range(meta),
+        def_kind
+      )
+
+    expr =
+      case expr do
+        nil ->
+          # function head
+          nil
+
+        [do: do_block] ->
+          # do block only
+          do_block
+
+        _ ->
+          if is_list(expr) and Keyword.has_key?(expr, :do) do
+            # do block with receive/catch/else/after
+            # wrap in try
+            # NOTE origin kind may be not correct here but origin is not used and
+            # elixir uses it only for error messages in elixir_clauses module
+            {:try, [{:origin, def_kind} | meta], [expr]}
+          else
+            # elixir raises here
+            expr
+          end
+      end
+
+    {_e_body, state, _env_for_expand} =
+      expand(expr, state, env_for_expand)
+
+    # restore vars from outer scope
+    state =
+      %{state | caller: false}
+
+    state =
+      unless has_unquotes do
+        # restore module vars
+        remove_func_vars_scope(state, state_orig)
+      else
+        # remove scope
+        remove_vars_scope(state, state_orig)
+      end
+
+    # result of def expansion is fa tuple
+    {{name, arity}, state, env}
+  end
+
+  defp expand_macro(
+         meta,
+         ExUnit.Case,
```

Comment:
Ideally we would find ways of making all ExUnit features work without special cases!